### ISSUE TITLE
Ability to configure the stroke width on a per icon basis

### ISSUE THREAD

**Issue by nickjj** [adamchainz/heroicons#24]
Hi,

Thanks for making this (I'm using it with Jinja 2).

I'm wondering if we can get an optional `stroke_width` parameter which defaults to 2 which is the same as Heroicons.

### Use case and examples

One thing that I tend to do quite often is tweak the `stroke-width` of an icon. This lets you create thin or bold icons.

Here's a few HTML examples:

```html
<div class="p-4 space-y-4">
  <svg xmlns="http://www.w3.org/2000/svg" class="h-10 w-10" fill="none" viewBox="0 0 24 24" stroke="currentColor">
    <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z" />
  </svg>

  <svg xmlns="http://www.w3.org/2000/svg" class="h-10 w-10" fill="none" viewBox="0 0 24 24" stroke="currentColor">
    <path stroke-linecap="round" stroke-linejoin="round" stroke-width="1" d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z" />
  </svg>

  <svg xmlns="http://www.w3.org/2000/svg" class="h-10 w-10" fill="none" viewBox="0 0 24 24" stroke="currentColor">
    <path stroke-linecap="round" stroke-linejoin="round" stroke-width="4" d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z" />
  </svg>
</div>
```

Which produces:

![image](https://user-images.githubusercontent.com/813219/122657181-23d8a200-d12f-11eb-80d6-b6b64c2a2b3e.png)

The only difference is the `stroke-width` is being set to 2 (default), 1 and 4. Every other attribute is the same.

You can play around with this example at: https://play.tailwindcss.com/625MHAeePO

**Comment by adamchainz** [comment]
This seems useful. Could you try making a PR?

**Comment by nickjj** [comment]
I tried before opening the issue but I couldn't see how to manipulate the properties of the `path`.

**Comment by adamchainz** [comment]
The library just reads the text of the icon files and spits it out into the HTML. Since the icon files are pretty predictable, I'd try a simple `str.replace()` call for `stroke-width`, which shouldn't appear anywhere else in the icon files. See https://github.com/adamchainz/heroicons/blob/a504d9043059cbf9d895980bf0aaa7fccb0bc5b7/src/heroicons/jinja.py#L21

**Comment by nickjj** [comment]
Thanks.

By the way, before going all-in with this library I had a question about `load_icon`. I see it's using `lru_cache` with a maxsize of 128. Would you mind walking through how this gets cached and what the implications are of calling this function with different arguments?

For example if you call this function 6 times with 6 different names, is it going to do 6 disk access operations where it opens the zip file? Basically every time you reference an icon it's going to do a very expensive operation until it's cached? But then the cache will get start to get recycled if you happen to use more than 128 icons?

**Comment by adamchainz** [comment]
The contents of each icon file are cached in memory.

The disk access operations will actually be pretty quick already - the operating system will typically cache some or all of the small zip file in memory on first access. The cache is only there to speed up reading commonly accessed icons further, especially when icons may be reused repeatedly on the same page.

Yes, the cache will fill up at 128 icons. That's a bit of an arbitrary cutoff but I'd estimate most sites don't use more than about that number very frequently. We could bump it up but there is definitely a tradeoff with keeping rare icon data always in memory, especially since the cache cannot be shared between web server processes.

**Comment by cachitas** [comment]
Following the idea of providing a value for `stroke-width`, I think we could think about setting any path property in a similar fashion, e.g. `vector-effect`, `stroke-linecap`, etc.

I encountered an [example](https://tailwindui.com/components/application-ui/feedback/empty-states#component-42930c785190e14c24b53ba822f4a92c) on TailwindUI where they set `vector-effect="non-scaling-stroke"` and it would be nice if we could accomplish the same using this tool.

I will leave the PR for @nickjj if he still wants to tag along, but would like to hear @adamchainz thoughts on where to go from here: 
1. Implement a generic way of setting any path property at will.
2. Implement just `stroke-width` as initially proposed.
3. Anything in between?

**Comment by adamchainz** [comment]
We could push certain attributes down to the path. It looks like that would be this list at current:

* `stroke-linecap`
* `stroke-linejoin`
* `stroke-width`
* `vector-effect`

Know of any others? I don't have TailwindUI so can't see the example(s).

**Comment by cachitas** [comment]
The examples usually reference heroicons without any modifications. On the rare occasions where they use custom SVGs, I only saw  those 4 properties set. I didn't go through all the examples though.

**Comment by nickjj** [comment]
I like the idea of allowing any attribute to be passed in. 

t's one of those things where it doesn't do any harm when you don't need or use it but when you need it, it's extremely useful because the only alternative is not to use this library for that 1 icon you want to customize.

**Comment by adamchainz** [comment]
I've implemented the "push down" of those four attributes to `<path>` elements, and released it in version 1.1.0. Give that a try!

There may be more attributes to add to the list in the future, from the [presentation attributes list on `<path>`](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/path) - we can add them when relevant.

**Comment by cachitas** [comment]
We may need to look into this again due to a fix on [heroicons 1.0.6](https://github.com/tailwindlabs/heroicons/releases/tag/v1.0.6).

**Comment by adamchainz** [comment]
Thank you for the heads-up @cachitas
